sdk/test-utils/perf/src/options.ts

```diff
@@ -53,6 +53,15 @@ export type PerfOptionDictionary<TOptions = Record<string, unknown>> = {
   [longName in keyof TOptions]: OptionDetails<TOptions[longName]>;
 };
 
+/**
+ * This is exactly same as {@link PerfOptionDictionary}, but the `value` is required.
+ * If it's absent and is required, we throw during validation.
+ */
+export type ParsedPerfOptions<TOptions = Record<string, unknown>> = {
+  [longName in keyof TOptions]: OptionDetails<TOptions[longName]> &
+    Pick<Required<OptionDetails<TOptions[longName]>>, "value">;
```

Comment:
Suppose the option is not required and there is no provided default value, then could the option be undefined? That would break this contract. Off the top of my head, that's possible, but not 100%.

Reply:
yeah, what about something like `& { value: TOptions[longName] }`?

Reply:
I think that has the same problem. Maybe `... & ({ required: true; value: TOptions[longName] } | { required: false | undefined })` would work (assuming an option not being required is the default).

Reply:
wait, in this case the idea is that value must be required 🤔

Reply:
Maybe I'm getting confused but the issue I see is that the value _could_ possibly be undefined (in the case of an option having `required: false`) when we cast it to `ParsedPerfOption`. So in that case we'd be making a false assertion that the value is not undefined.

Reply:
let’s break this down (at least, to share what I’m thinking),

- `TOptions[longName]` is the type of the `longName` property on the `TOptions` type. Let’s call it `_OptionType` for this analysis.
- `Required<OptionDetails<_OptionType>>` will extract only the required properties out of `OptionDetails<_OptionType> `.
- Then, `Pick` will only extract the `value` property.

Since the `value` property on `OptionDetails` is the parametrized `TType`, then the above should be equivalent to `{ value: TOptions[longName] }`.

It’s a simplification, I believe. We could start thinking based on a simpler version of the type.

Reply:
@timovv ok ok. I thought this type was only strictly used with object that had value properties with `required: true`.

Reply:
Yeah, if you see the code here, value should always be defined: https://github.com/Azure/azure-sdk-for-js/blob/e4b27351fe26519c91433c020b95215e86a28f7e/sdk/test-utils/perf/src/options.ts#L177 — I could be missing something.

Reply:
I agree that your simplification is equivalent to what's already there (and is probably a good simplification regardless) but am not convinced it addresses the issue I had in mind.

Reply:
yeah I don’t think I addressed your concerns 😆 I’m still trying to water this down. It’s a bit confusing

Reply:
https://github.com/Azure/azure-sdk-for-js/blob/e4b27351fe26519c91433c020b95215e86a28f7e/sdk/test-utils/perf/src/options.ts#L170-L172

If `required` is `false`, we won't throw if `value` is `undefined` and so it could still be undefined if we get to that point.

Reply:
so, even with my simplification, the type of `value` essentially gets assigned as the type of the last property read by `keyof` 🤔

Reply:
@timovv I think that, to solve your concern, either:

- Make `parsePerfOption` more strict.
- Make `parsePerfOption` return `PerfOptionDictionary<TOptions>`.

🤔 what do you think?

Reply:
What about this?

```ts
type ParsedPerfOptions<TValue = string, TOptions = { value: TValue, [key: string]: unknown; }> = {
  [longName in keyof TOptions]: OptionDetails<TOptions[longName]>
} & { value: OptionDetails<TValue> & { required: true } };

const parsedOptions: ParsedPerfOptions = {
  value: {
    required: true, // errors if false
    value: "hola", // errors if not string
    description: "salutations"
  }
}
```

Reply:
@timovv We could do something like what you’re describing: `| { required: false /* etc */`, but is it the intention?

Please let me know if I’m wrong 🙏

Reply:
Agreed on the first option. I think we should make `parsePerfOption` more strict by requiring (and setting) a default value if `required` is set to false. That said, it might be desirable sometimes to have an option be `undefined` by default when it's not required -- but in that case we could probably set the default to `null` instead and get basically the same behavior...

Your second option is, IIRC, the status quo before this PR. The reason I think this change is in here in the first place (correct me if I'm wrong, @HarshaNalluru) is that it's annoying -- and actually a lint error now that we reenabled the linter perf tests -- to have non-null assertions on the `value` property scattered all over the place, which is what we were doing before.

The only reason I brought this issue up in the first place is that I was playing around with fixing this up in an earlier PR. but I didn't tackle it in the end because I got really confused in this area and decided it wasn't worth it at the time! But if the linter is becoming a pain now, it's probably worth addressing, since it will only get more annoying as time goes on 😄

Reply:
We could make it throw if we can’t derive a value from either the user’s input or the default value. If neither of those is present, then throwing seems safe to me. Keeps things simple.

Reply:
> @timovv We could do something like what you’re describing: | { required: false /* etc */, but is it the intention?

Yeah, so the intention behind my idea (can't say I've tested it) is that in cases when `required` is set to true, you wouldn't have to make that non-null assertion when you grab the value. But when `required` is false, you would have to make that assertion (which makes sense, because it might not be there, since it's optional to provide it).

Reply:
We can also default to null in the code I shared above:

```ts
type ParsedPerfOptions<TValue = null, TOptions = { value: TValue, [key: string]: unknown; }> = {
  [longName in keyof TOptions]: OptionDetails<TOptions[longName]>
} & { value: OptionDetails<TValue> & { required: true } };

const parsedOptions: ParsedPerfOptions<string> = {
  value: {
    required: true, // errors if false
    value: "hola", // errors if not string
    description: "salutations"
  }
}
```

Reply:
I have shared a lot of spam so far 😆

I’ll leave this up to you both, @HarshaNalluru and @timovv 🤘

Reply:
omg, I was away for a couple of hours and see.. 😂
I like the suggestion, I'll test and incorporate it.

Reply:
I like the enthusiasm 😍

Reply:
I can't find a better approach with the conditional types/mapped types without breaking the existing APIs or interfaces.
The main problem is TS treats `value ?: T` and `value : T | undefined` the same.

Rethinking @timovv's question..
I don't think anybody is affected with what I have in this PR.
I would suggest that we don't consider `undefined` as an option. (Just because none of our ton of perf tests do.)
This whole discussion is out of scope for this PR anyways.

Having said that, 
- let's go with what I have
- let's go with what I have and then revisit when we handle https://github.com/Azure/azure-sdk-for-js/issues/18032
- let's not change anything in this PR (and suffer with the lint errors 😛)

Reply:
No, if we are breaking APIs, it'll break all the existing perf tests, handling them would be out of scope for this PR.
We can consider breaking the APIs, but that should only be part of a future PR.

Reply:
Let’s make an issue 🙆‍♂️

Reply:
IMO Harsha's right that this would tie in nicely with the work in https://github.com/Azure/azure-sdk-for-js/issues/18032. I have a few ideas for refactoring how options work (it's a bit funky/limiting how the options definition is an instance member of `PerfTest`, for starters) so may be worth having a look at all of it in one go.

Reply:
ok ok I have added a comment here: https://github.com/Azure/azure-sdk-for-js/issues/18032#issuecomment-983038502

However, I don’t think that my proposal is a breaking change, since it should only break if there are options passed with `value: { required: false }`. Harsha and I will meet today to discuss this. Regardless, I am confident this can be solved via #18032 👍

Reply:
So, turns out my previous design [here](https://github.com/Azure/azure-sdk-for-js/pull/18811#discussion_r756508020) was way far from the goal! 😆 but this one actually solved the problem: https://github.com/Azure/azure-sdk-for-js/pull/18811#discussion_r756494838 🎉

Reply:
Final type
```ts
export type ParsedPerfOptions<TOptions = Record<string, unknown>> = {
  [longName in keyof TOptions]: OptionDetails<TOptions[longName]> & { value: TOptions[longName] };
};
```